### PR TITLE
Ruff, Github Actions and other refactoring

### PR DESCRIPTION
**Pull request by FBruzzesi**
# Description

This is another large PR I guess, here the list of changes:

- Migrate from setup.py to pyproject.toml
  - keeping setup.py just to distribute using the same process but could even be removed
- Move away from black and flake8 to adopt ruff both as linter and formatter (solve #601)
- Cleanup some github actions/workflows
- Every `sklego/*.py` and `tests/*.py` edit is due to ruff
- Only exception being `sklego/__init__.py` 

# Checklist:

- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/dependabot.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+version: 2
```
Comment (FBruzzesi) [inline .github/dependabot.yaml:1]: Periodically checks for github action updates - are we interested in this?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  pull_request:
+    branches:
+    - main
+
+jobs:
+  test:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
```
Comment (FBruzzesi) [inline .github/workflows/test.yml:13]: Moved all os into a single workflow

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        python-version: [
+          # "3.6", Unsupported on ubuntu-latest (22.04)
+          # "3.7", Long series of scikit-learn failures
+          "3.8",
+          "3.9",
+          "3.10",
+          "3.11",
+          # "3.12", # Numba not available for Python 3.12: https://github.com/numba/numba/issues/9197
```
Comment (FBruzzesi) [inline .github/workflows/test.yml:21]: Non-working version are commented with their reason for breaking

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    "scikit-learn>=1.0",
+    "importlib-metadata >= 1.0; python_version < '3.8'",
+]
+
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
```
Comment (FBruzzesi) [inline pyproject.toml:34]: Tagged only the version for which workflows run without errors

File: sklego/__init__.py [hunk 5]
```diff
@@ -1 +1,9 @@
-__version__ = "0.7.0"
+import sys
+
+if sys.version_info >= (3, 8):
+    from importlib import metadata
+else:
+    import importlib_metadata as metadata
+
+__title__ = "sklego"
+__version__ = metadata.version("scikit-lego")
```
Comment (FBruzzesi) [inline sklego/__init__.py:9]: This took me sooo long to debug due to the different naming between distribution and module

File: sklego/decomposition/__init__.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-from .pca_reconstruction import PCAOutlierDetection
-from .umap_reconstruction import UMAPOutlierDetection
+from sklego.decomposition.pca_reconstruction import PCAOutlierDetection
+from sklego.decomposition.umap_reconstruction import UMAPOutlierDetection
```
Comment (FBruzzesi) [inline sklego/decomposition/__init__.py:2]: I made all imports absolute, maybe unnecessary

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+license = {file = "LICENSE"}
+readme = "readme.md"
+requires-python = ">=3.6"
+authors = [
+    {name = "Vincent D. Warmerdam"},
+    {name = "Matthijs Brouns"},
+]
+
+maintainers = [
+    {name = "Francesco Bruzzesi"}
```
Comment (FBruzzesi) [inline pyproject.toml:19]: Kinda sneak myself in there 😁

File: .github/workflows/test.yml (hunk 3 above)
Comment (koaning) [reply .github/workflows/test.yml:21]: It think it's fine to remove both 3.6 and 3.7 at this point though. Both have reached end of life officially.

File: .github/dependabot.yaml (hunk 1 above)
Comment (koaning) [reply .github/dependabot.yaml:1]: I'm willing to try it. But I'm in favor of removing it the moment it starts producing noise.

File: pyproject.toml (hunk 7 above)
Comment (koaning) [reply pyproject.toml:19]: Rightly so!

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    "matplotlib>=3.0.2",
+    "jupyter>=1.0.0",
+    "jupyterlab>=0.35.4",
+]
+
+dev = [
+    "scikit-lego[all,test,docs]",
+    "pre-commit>=1.18.3",
+    "ruff>=0.1.6",
+    "hatch",
```
Comment (koaning) [inline pyproject.toml:81]: I might be missing something, but where are we using hatch? I don't think I mind, but we should add a minimum version I guess.

**Review by koaning — COMMENTED** [review]
I left some minor comments and a question about hatch. These feel relatively minor though so I'm assuming I'll just merge it once the answers ping back. 

Good work though, this feels like a pretty big cleanup!

File: sklego/__init__.py (hunk 5 above)
Comment (koaning) [reply sklego/__init__.py:9]: Yeah, that's certainly a bit awkward.

File: pyproject.toml (hunk 8 above)
Comment (FBruzzesi) [reply pyproject.toml:81]: Oh nice catch! I was trying to completely move away from using setup.py using another building tool. For some reason I couldn't fully manage to do so, but forgot to remove `hatch`.

I will fix it as soon as I log into a laptop 😁

File: .github/dependabot.yaml (hunk 1 above)
Comment (FBruzzesi) [reply .github/dependabot.yaml:1]: In my experience it shouldn't be too noisy as we are not using very fast changing github actions

**Review by koaning — APPROVED** [review]

**Pull request merged**